src/ingestion/sources/bigquery.py

```diff
@@ -0,0 +1,200 @@
+    def _text_int_rate(cls):
+        return "SUM(ARRAY_LENGTH(REGEXP_EXTRACT_ALL(CAST({} AS STRING), r'^([-+]?[0-9]+)$'))) / CAST(COUNT(*) AS NUMERIC)"
+
+    @classmethod
+    def _text_number_rate(cls):
+        return "SUM(ARRAY_LENGTH(REGEXP_EXTRACT_ALL(CAST({} AS STRING), r'^[\+\-]?\d*\.?\d+(?:[Ee][\+\-]?\d+)?$'))) / CAST(COUNT(*) AS NUMERIC)"
+
+    @classmethod
+    def _text_uuid_rate(cls):
+        return "SUM(ARRAY_LENGTH(REGEXP_EXTRACT_ALL(CAST({} AS STRING2), r'^([0-9a-fA-F]{{8}}-[0-9a-fA-F]{{4}}-[0-9a-fA-F]{{4}}-[0-9a-fA-F]{{4}}-[0-9a-fA-F]{{12}})$'))) / CAST(COUNT(*) AS NUMERIC)"
+
+    @classmethod
+    def _text_all_spaces_rate(cls):
+        return "SUM(ARRAY_LENGTH(REGEXP_EXTRACT_ALL(CAST({} AS STRING), r'^\s*$'))) / CAST(COUNT(*) AS NUMERIC)"
+
+    @classmethod
+    def _text_null_keyword_rate(cls):
+        return "SUM(CASE WHEN UPPER(CAST({} as STRING)) IS NULL THEN 1 ELSE 0 END) / CAST(COUNT(*) AS NUMERIC)"
+
+    @classmethod
+    def _zero_rate(cls): # TODO: ?
+        return "SUM(CASE WHEN UPPER(CAST({} as STRING)) IS NULL THEN 1 ELSE 0 END) / CAST(COUNT(*) AS NUMERIC)"
+
+    @classmethod
+    def _negative_rate(cls):
+        return "SUM(CASE WHEN {} < 0 THEN 1 ELSE 0 END) / CAST(COUNT(*) AS NUMERIC)"
+
+    @classmethod
+    def _completeness(cls):
+        return "COUNT({}) / CAST(COUNT(*) AS NUMERIC)"
+    
+    @classmethod
+    def _freshness(cls):
+        return "TIMESTAMP_DIFF(MAX({}), CURRENT_TIMESTAMP, MINUTE)"
+
+    @classmethod
+    def schema_tables_query(cls, database_name, schema_name):
+        raise NotImplementedError
+
+    @classmethod
+    def table_metrics_query(cls, monitor, discovery_data, minutes_ago):
+        builder = BigQueryMetricsQueryBuilder(cls, monitor, discovery_data, minutes_ago)
+        query = builder.compile()
+        return query
+
+
+    @classmethod
+    def schema_columns_query(cls, database_name, schema_name):
+        return """
+        SELECT lower(c.table_name) AS `NAME`,
+            lower(c.column_name) AS `COL_NAME`,
+            lower(c.data_type) AS `COL_TYPE`,
+            c.ordinal_position AS `COL_SORT_ORDER`,
+            lower(c.table_catalog) AS `DATABASE`,
+            lower(c.table_schema) AS `SCHEMA`,
+            CASE
+                lower(t.table_type)
+                WHEN 'view' THEN 'true'
+                ELSE 'false'
+            END `IS_VIEW`
+        FROM {schema_name}.INFORMATION_SCHEMA.COLUMNS AS c
+            LEFT JOIN {schema_name}.INFORMATION_SCHEMA.TABLES t ON c.TABLE_NAME = t.TABLE_NAME
+            AND c.TABLE_SCHEMA = t.TABLE_SCHEMA
+        WHERE LOWER(c.table_schema) = LOWER('{schema_name}')
+            AND LOWER(c.table_catalog) = LOWER('{database_name}')
+        """.format(database_name=database_name, schema_name=schema_name)
+
+    @classmethod
+    def query_access_logs_query(cls):
+        raise NotImplementedError
+        
+    @classmethod
+    def query_copy_logs_query(cls):
+        raise NotImplementedError
```

Comment:
theoretically should have notimplemented on base and let it fall through and move implementations to corresponding ones that work

Reply:
sounds good, removing